File: FastCompile/Python/jsonrepl.py
#!/usr/bin/env python
import json
import traceback
import sys
import imp
from cStringIO import StringIO

def resp(result):
    result["backendVersion"] = "one:python:jsonrepl:20180122"
    print json.dumps(result)

while True:
    try:
        requestLine = raw_input()
    except:
        break

    try:
        request = json.loads(requestLine)
        if request["cmd"] == "compile":
            original_stdout = sys.stdout
            sys.stdout = result_stdout = StringIO()
            try:
                sys.modules["one"] = imp.new_module("one")
                exec(request["stdlibCode"], sys.modules["one"].__dict__)
                exec(request["code"], {})
            finally:
                sys.stdout = original_stdout

            resp({ "result": result_stdout.getvalue() })
        else:
            resp({ "errorCode": "unknown_command", "exceptionText": "Unknown command: " + request["cmd"] })
    except Exception as e:
        resp({ "errorCode": "unexpected_exception", "exceptionText": traceback.format_exc() })

File: FastCompile/Ruby/jsonrepl.rb
require 'json'
require 'stringio'

trap 'INT' do exit end
$stdout.sync = true
$realStdout = $stdout
    
def resp(result)
    result["backendVersion"] = "one:ruby:jsonrepl:20180122"
    $realStdout.puts(JSON.generate(result))
end

while requestLine = gets
    begin
        $stdout = StringIO.new
        request = JSON.parse(requestLine)
        eval request['stdlibCode']
        result = eval request['code'].sub(/require 'one'/, "")
        resp({ :result => $stdout.string })
    rescue Exception
        resp({ :exceptionText => "#{$@.first}: #{$!.message} (#{$!.class})" + $@.drop(1).map{|s| "\n\t#{s}"}.join("") })
    ensure
        $stdout = $realStdout
    end
end
